lib/simple-seeder.js

```diff
@@ -0,0 +1,221 @@
+      release()
+    }
+  }
+
+  async _add (key, opts) {
+    if (this.destroying) return
+    if (this.has(key)) throw new Error('Resource already added')
+
+    const publicKey = HypercoreId.decode(key)
+    const id = HypercoreId.encode(publicKey)
+
+    // Maybe all this belongs to a "Resource" class, including the internals _createCore/Bee/etc
+    const info = {
+      key: id,
+      type: opts.type,
+      instance: null,
+      description: opts.description || null,
+      seeders: null,
+      discovery: null,
+      blocks: { down: speedometer(), up: speedometer() },
+      network: { down: speedometer(), up: speedometer() },
+      external: opts.external !== false
+    }
+
+    if (opts.type === 'core') info.instance = await this._createCore(publicKey, info)
+    else if (opts.type === 'bee') info.instance = await this._createBee(publicKey, info)
+    else if (opts.type === 'drive') info.instance = await this._createDrive(publicKey, info, opts)
+    else if (opts.type === 'list') info.instance = await this._createBee(publicKey, info, { keyEncoding: 'utf-8', valueEncoding: 'json' })
+    else if (opts.type === 'seeders') {
+      info.type = 'drive'
+      info.instance = await this._createDrive(publicKey, info, { seeders: true })
+    } else throw new Error('Type is required')
+
+    this.resources.set(id, info)
+
+    return info
+  }
+
+  async put (key, opts) {
+    const release = await this.lock()
+    try {
+      return await this._put(key, opts)
+    } finally {
+      release()
+    }
+  }
+
+  async _put (key, opts) {
+    if (!this.has(key)) return this._add(key, opts)
+
+    const info = this.get(key)
+
+    if (info.type !== opts.type) {
+      await this._remove(key)
+      return this._add(key, opts)
+    }
+
+    if (!!info.seeders !== opts.seeders) {
+      if (opts.seeders) {
+        if (info.seeders) throw new Error('Seeders was already enabled') // Should not happen
+        const publicKey = HypercoreId.decode(info.key)
+        info.seeders = await this._createSeeders(publicKey)
+      } else {
+        await info.seeders.destroy()
+        info.seeders = null
+      }
+    }
+
+    info.description = opts.description || null
+
+    return info
+  }
+
+  async remove (key) {
+    const release = await this.lock()
+    try {
+      return await this._remove(key)
+    } finally {
+      release()
+    }
+  }
+
+  async _remove (key) {
+    const info = this.get(key)
+
+    if (info.seeders) await info.seeders.destroy()
+    if (info.discovery) await info.discovery.destroy()
+    await info.instance.close()
+
+    this.resources.delete(info.key)
+  }
+
+  has (key) {
+    const id = HypercoreId.normalize(key)
+    return this.resources.has(id)
+  }
+
+  get (key) {
+    const id = HypercoreId.normalize(key)
+    const info = this.resources.get(id)
+    if (!info) throw new Error('Resource does not exists')
+    return info
+  }
+
+  filter (onfilter) {
+    const arr = []
+    for (const info of this.values()) {
```

Comment:
I think you can just use .filter() here, like 
`return [...this.values()].filter(onfilter)

Reply:
Functional programming is illegal you know